### PR TITLE
Apply custom log levels to Truffle logger

### PR DESCRIPTION
**Pull request by hubertp**
### Pull Request Description

Previously custom log levels applied only to non-Truffle loggers. To allow it, filtering has to be applied appropriately at two places - first at Java's Handler and then essentially re-confirmed at SLF4J's logger to which the former forwards to.
Filters compose in an `AND` condition, therefore default log level check had to be merged into our custom filters.

`TruffleLogger` has a builtin functionality to perform the filtering when context is configured appropriately. This should be much more efficient than adding a `Filter` to the JUL Handler explicitly.

### Important Notes

```
JAVA_OPTS="-org.enso.compiler.SerializationManager.Logger.level=debug" ./built-distribution/enso-engine-0.0.0-dev-linux-amd64/enso-0.0.0-dev/bin/enso  --run
```
will now assign a custom log level to `SerializationManager` Logger.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.

### CODE REVIEW

**Review by JaroslavTulach — CHANGES_REQUESTED** [review]
Can we write a benchmark that is using `TruffleLogger` first and then measure whether it is still _no overhead_ when disabled? If so, then I am OK with the change.

However I doubt that's going to be the case. Using `Filter` approach must in my opinion have an overhead. The only way to eliminate it completely is to use `Assumption` as the `TruffleLogger` infrastructure does.

File: engine/runner/src/main/java/org/enso/runner/LogLevelFilter.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+import java.util.AbstractMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.logging.Level;
+import java.util.stream.Collectors;
+import org.enso.logger.Converter;
+import org.enso.logger.LevelAndNameFilter;
+import org.enso.logger.config.LoggersLevels;
+
+public class LogLevelFilter implements LevelAndNameFilter {
```
Comment (JaroslavTulach) [inline engine/runner/src/main/java/org/enso/runner/LogLevelFilter.java]: `final`, I guess.

File: engine/runner/src/main/scala/org/enso/runner/ContextFactory.scala [hunk 2]
```diff
@@ -95,11 +105,26 @@ class ContextFactory {
...
+      builder.option(
         RuntimeOptions.LOG_LEVEL,
         logLevelName
       )
-      .logHandler(JulHandler.get())
+    } else {
+      builder.option(
+        RuntimeOptions.LOG_LEVEL,
+        java.util.logging.Level.ALL.getName
+      )
```
Comment (JaroslavTulach) [inline engine/runner/src/main/scala/org/enso/runner/ContextFactory.scala]: I believe we have to enumerate all the `enso....Logger.level` properties and apply them one by one with `log.enso...` prefix.

**Comment by hubertp** [comment]
> Can we write a benchmark that is using `TruffleLogger` first and then measure whether it is still _no overhead_ when disabled? If so, then I am OK with the change.
> 
> However I doubt that's going to be the case. Using `Filter` approach must in my opinion have an overhead. The only way to eliminate it completely is to use `Assumption` as the `TruffleLogger` infrastructure does.

Thanks for the pointer on using `TruffleLogger` builtin filtering ability, I somehow missed that. It fitted our use-case perfectly.

File: lib/scala/logging-config/src/main/java/org/enso/logger/config/LoggersLevels.java [hunk 3]
```diff
@@ -55,6 +55,15 @@ public boolean isEmpty() {
     return loggers.isEmpty();
   }
 
+  /**
+   * Checks if any of the custom levels refer to Enso's loggers.
+   *
+   * @return true if any of the custom log levels are for `org.enso.*` class
+   */
+  public boolean hasEnsoLoggers() {
```
Comment (JaroslavTulach) [inline lib/scala/logging-config/src/main/java/org/enso/logger/config/LoggersLevels.java:63]: Why check for `hasEnsoLoggers`? Don't we want to enabled for example also JavaScript or Python loggers dynamically?

**Review by JaroslavTulach — APPROVED** [review]

File: lib/scala/logging-config/src/main/java/org/enso/logger/config/LoggersLevels.java (hunk 3 above)
Comment (hubertp) [reply lib/scala/logging-config/src/main/java/org/enso/logger/config/LoggersLevels.java:63]: I didn't think of that case, TBH. Can you give me an example?

Comment (JaroslavTulach) [reply lib/scala/logging-config/src/main/java/org/enso/logger/config/LoggersLevels.java:63]: I haven't found any use of `TruffleLogger` in JavaScript. GraalPython seems to be using [TruffleLogger somehow](https://github.com/oracle/graalpython/blob/a9b6cd7edc6e25f9b1d4011316e2d153ee88150f/graalpython/com.oracle.graal.python/src/com/oracle/graal/python/PythonLanguage.java#L291).

Comment (Akirathan) [reply lib/scala/logging-config/src/main/java/org/enso/logger/config/LoggersLevels.java:63]: `regex` uses TruffleLoggers for sure. And they are pretty verbose.

Comment (Akirathan) [reply lib/scala/logging-config/src/main/java/org/enso/logger/config/LoggersLevels.java:63]: I am not sure if I fully understand this discussion. But if this change means that we cannot suppress, for example, `org.truffle.r.*.logger` or `org.graalpython*.logger`, then that is unfortunate.

**Pull request merged**